jsonrpc/ipc_test.go

```diff
@@ -0,0 +1,249 @@
+		return fmt.Errorf("recv msg does not match. got %v, want %v", string(buffer), want)
+	}
+	return nil
+}
+
+// The caller is responsible for closing the server.
+func testUnix(t *testing.T, method jsonrpc.Method, listener jsonrpc.EventListener, endpoint string) *jsonrpc.Ipc {
+	rpc := jsonrpc.NewServer(1, utils.NewNopZapLogger()).WithListener(listener)
+	require.NoError(t, rpc.RegisterMethods(method))
+	ipc, err := jsonrpc.NewIpc(rpc, endpoint, utils.NewNopZapLogger())
+	require.NoError(t, err)
+	return ipc
+}
+
+func TestIpcHandler(t *testing.T) {
+	var (
+		msg    = `{"jsonrpc" : "2.0", "method" : "test_echo", "params" : [ "abc123" ], "id" : 1}`
+		want   = `{"jsonrpc":"2.0","result":"abc123","id":1}`
+		method = jsonrpc.Method{
+			Name:   "test_echo",
+			Params: []jsonrpc.Parameter{{Name: "msg"}},
+			Handler: func(msg string) (string, *jsonrpc.Error) {
+				return msg, nil
+			},
+		}
+	)
+
+	t.Run("socket cleanup", func(t *testing.T) {
+		var (
+			ep          = path.Join(t.TempDir(), "juno.ipc")
+			ctx, cancel = context.WithCancel(context.Background())
+			srvWg       sync.WaitGroup
+		)
+		srvWg.Add(1)
+
+		defer cancel()
+		listener := CountingEventListener{}
+		srv := testUnix(t, method, &listener, ep)
+		go func() {
+			assert.NoError(t, srv.Run(ctx))
+			srvWg.Done()
+		}()
+
+		_, err := new(net.Dialer).DialContext(context.Background(), "unix", ep)
+		assert.NoError(t, err)
+		cancel()
+		srvWg.Wait()
+		_, err = new(net.Dialer).DialContext(context.Background(), "unix", ep)
+		assert.ErrorIs(t, err, os.ErrNotExist)
+	})
+
+	t.Run("single conn", func(t *testing.T) {
+		var (
+			ep          = path.Join(t.TempDir(), "juno.ipc")
+			ctx, cancel = context.WithCancel(context.Background())
+			srvWg       sync.WaitGroup
+		)
+		srvWg.Add(1)
+		defer func() {
+			cancel()
+			srvWg.Wait()
+		}()
+
+		listener := CountingEventListener{}
+		srv := testUnix(t, method, &listener, ep)
+		go func() {
+			assert.NoError(t, srv.Run(ctx))
+			srvWg.Done()
+		}()
+
+		conn, err := new(net.Dialer).DialContext(context.Background(), "unix", ep)
+		assert.NoError(t, err)
+		assert.NoError(t, exchangeMsg(conn, msg, want))
+	})
+
+	t.Run("multiple conns", func(t *testing.T) {
+		var (
+			items       = 1024
+			conns       = make([]net.Conn, items)
+			errCh       = make(chan error, items)
+			ep          = path.Join(t.TempDir(), "juno.ipc")
+			ctx, cancel = context.WithCancel(context.Background())
+			handlerWg   sync.WaitGroup
+		)
+		handlerWg.Add(1)
+		defer func() {
+			cancel()
+			handlerWg.Wait()
+		}()
+
+		listener := CountingEventListener{}
+		srv := testUnix(t, method, &listener, ep)
+		go func() {
+			assert.NoError(t, srv.Run(ctx))
+			handlerWg.Done()
+		}()
+
+		for i := 0; i < items; i++ {
+			conn, err := new(net.Dialer).DialContext(ctx, "unix", ep)
+			assert.NoError(t, err)
+			conns[i] = conn
+			go func() {
+				errCh <- exchangeMsg(conn, msg, want)
+			}()
+		}
+
+		for i := 0; i < items; i++ {
```

Comment:
~Do we need the for loop?~ I think I get it now. `exchangeMsg` will return `nil` on `errCh` for each item. Is that right?

Reply:
Yes. It's ping-pong test for multiple connections.